src/monitor/visual_graph.rs

```diff
@@ -0,0 +1,126 @@
+//! Intermediate representation of a circuit graph suitable for
+//! conversion to a visual format like dot.
+
+type Id = String;
+
+/// Visual representation of a circuit graph.
+///
+/// The graph consists of a tree of cluster nodes populated with simple nodes.
+pub struct Graph {
+    nodes: ClusterNode,
+    edges: Vec<Edge>,
+}
+
+impl Graph {
+    pub(super) fn new(nodes: ClusterNode, edges: Vec<Edge>) -> Self {
+        Self { nodes, edges }
+    }
+
+    /// Convert graph to dot format.
+    pub fn to_dot(&self) -> String {
+        let mut lines: Vec<String> = vec!["digraph {".to_string(), "node [shape=box]".to_string()];
+        lines.append(&mut self.nodes.to_dot());
+        for edge in self.edges.iter() {
+            lines.push(edge.to_dot());
+        }
+        lines.push("}".to_string());
+        lines.join("\n")
+    }
+}
+
+pub(super) struct SimpleNode {
+    id: Id,
+    label: String,
+}
+
+impl SimpleNode {
+    pub(super) fn new(id: Id, label: String) -> Self {
+        Self { id, label }
+    }
+
+    fn to_dot(&self) -> Vec<String> {
+        vec![format!("{}[label=\"{}\"]", self.id, self.label)]
+    }
+}
+
+/// A cluster node represents a subcircuit or a region.
+// TODO:
+// * Visually distinguish subcircuits from regions (e.g., dashed vs solid
+//   boundaries).
+pub(super) struct ClusterNode {
+    id: Id,
+    label: String,
+    nodes: Vec<Node>,
+}
+
+impl ClusterNode {
+    pub(super) fn new(id: Id, label: String, nodes: Vec<Node>) -> Self {
+        Self { id, label, nodes }
+    }
+
+    // TODO: We add a pair of enter/exit nodes to each cluster and connect all
```

Comment:
graphviz supports edges to nodes in a cluster.

Reply:
https://graphviz.org/Gallery/directed/cluster.html

Reply:
Not really. An edge must point to a node in the cluster.